File: week4/java/SudokuSolver_37.java
class Solution {
    public void solveSudoku(char[][] board) {
        if (board == null || board.length != 9 || board[0].length != 9) return;
        findSolution(board);
    }
    
    private boolean findSolution(char[][] board) {
        for (int i = 0; i < board.length; i++) {
            // verify given board
            if (board[i].length != 9) return false;
            
            for (int j = 0; j < board[i].length; j++) {
                if (board[i][j] != '.') continue;
                
                for (char c = '1'; c <= '9'; c++) {
                    if (!isValid(board, c, i, j)) continue;
                    
                    board[i][j] = c;
                    if (findSolution(board)) 
                        return true;
                    else 
                        board[i][j] = '.';    
                }
                return false;
            }
        }
        return true;
    }
    
    private boolean isValid(char[][] board, char c, int row, int col) {
        for (int i = 0; i < 9; i++) {
            // check row
            if (board[row][i] == c) return false;
            // check col
            if (board[i][col] == c) return false;
            // check sub square
            if (board[3 * (row / 3) + i / 3][3 * (col / 3) + i % 3] == c) return false;
        }
        return true;
    }
}

File: week4/python/SudokuSolver_37.py
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        length = len(board)
        rows = [0] * len(board)
        cols = [0] * len(board)
        squares = [0] * len(board)
        
        def isValid(x: int, y: int, num: str) -> bool:
            # 1st approach
            # check by iterations
#             for i in range(len(board)):
#                 if board[i][y] == num:
#                     return False
#                 if board[x][i] == num:
#                     return False
#                 if board[3 * (x // 3) + i // 3][3 * (y // 3) + i % 3] == num:
#                     return False
            
#             return True

            # 2nd approach
            # check by recording in bits
            mark = 1<<int(num)
            return (not bool(rows[x] & mark)) \
                    and (not bool(cols[y] & mark)) \
                    and (not bool(squares[3*(x//3) + y//3] & mark))
        
        
        def solve(r: int, c: int) -> bool:
            if r >= length:
                return True
            
            if c >= length:
                return solve(r + 1, 0)
            
            # check next if not a space
            if board[r][c] != '.':
                return solve(r, c + 1)            

            for z in range(1, 10):
                num = str(z)
                if not isValid(r, c, num):
                    continue

                board[r][c] = str(num)
                mark = 1<<int(board[r][c])
                rows[r] |= mark
                cols[c] |= mark
                squares[3*(r//3) + c//3] |= mark
                
                if solve(r, c + 1):
                    return True
                
                board[r][c] = '.'
                rows[r] ^= mark
                cols[c] ^= mark
                squares[3*(r//3) + c//3] ^= mark

            return False
                    
        for r in range(length):
            for c in range(length):
                if board[r][c] == '.':
                    continue
                
                mark = 1<<int(board[r][c])
                rows[r] |= mark
                cols[c] |= mark
                squares[3*(r//3) + c//3] |= mark
                
        solve(0,0)
